applications/affine-finetune-embeddings/main.py

```diff
@@ -0,0 +1,171 @@
+import optuna
+from pytorch_lightning.loggers import TensorBoardLogger
+from pytorch_lightning.callbacks import ModelCheckpoint
+import torch
+from torch.utils.data import DataLoader
+from dataset import EmbeddingDataset, load_and_split_data
+from model import SimilarityModel
+import pytorch_lightning as pl
+
+model_id = "BAAI/bge-small-en-v1.5"
+
+
+class Sampler(object):
+    """Base class for all Samplers.
+
+    Every Sampler subclass has to provide an __iter__ method, providing a way
+    to iterate over indices of dataset elements, and a __len__ method that
+    returns the length of the returned iterators.
+    """
+
+    def __init__(self, data_source):
+        pass
+
+    def __iter__(self):
+        raise NotImplementedError
+
+    def __len__(self):
+        raise NotImplementedError
+
+
+class StratifiedSampler(Sampler):
+    """Stratified Sampling
+
+    Provides equal representation of target classes in each batch
+    """
+
+    def __init__(self, class_vector, batch_size):
+        """
+        Arguments
+        ---------
+        class_vector : torch tensor
+            a vector of class labels
+        batch_size : integer
+            batch_size
+        """
+        self.n_splits = int(class_vector.size(0) / batch_size)
+        self.class_vector = class_vector
+
+    def gen_sample_array(self):
+        from sklearn.model_selection import StratifiedShuffleSplit
+        import numpy as np
+
+        s = StratifiedShuffleSplit(n_splits=self.n_splits, test_size=0.5)
+        X = torch.randn(self.class_vector.size(0), 2).numpy()
+        y = self.class_vector.numpy()
+        s.get_n_splits(X, y)
+
+        train_index, test_index = next(s.split(X, y))
+        return np.hstack([train_index, test_index])
+
+    def __iter__(self):
+        return iter(self.gen_sample_array())
+
+    def __len__(self):
+        return len(self.class_vector)
+
+
+def objective(trial):
+    # Hyperparameters to be optimized
+    embedding_size = 384
```

Comment:
this is a function of the model right?

Reply:
Yea bge small is 384 param. I hardcoded for now but I can change it

Reply:
lets me a config as just min, max embedding size